### ISSUE TITLE
Endianess issue : save/read game

### ISSUE THREAD

**Issue by BeWorld2018** [dxx-rebirth/dxx-rebirth#769]
Platform : MorphOS (PPC= / Big Endian

Hi, i try to build rxx to MorphOS platform but i found some endianess problem with save/load game on my platform.

https://github.com/dxx-rebirth/dxx-rebirth/blob/master/similar/main/state.cpp#L1281

Example :
Here : https://github.com/dxx-rebirth/dxx-rebirth/blob/master/similar/main/state.cpp#L1328 need to be SWAPINT()
I try to fix all but there are too much varible to fix, maybe you can do something.

I test to load from Windows saved game and working fine.

Thanks for your help.

**Comment by vLKp** [comment]
I disagree with your assertion.  For historical reasons, many parts of the save game are written in host endian, rather than explicitly big endian or explicitly little endian.  I think the loader would have been simpler if the save game was written in a fixed endian, but it is many years too late to fix that now.  As a result, the loader reads in the version field, checks whether the version appears to be foreign endian and, if so, makes a note to byte-swap all later multibyte values as they are loaded, which should result in them being converted to host endian.

Adding `SWAPINT` as you describe would cause little endian systems to write savegames that claim to be big endian, and vice versa.  I expect this would cause widespread problems loading such games.

Instead of describing a proposed change, please explain what you saw that led you to think there is an endianness problem.

**Comment by BeWorld2018** [comment]
ok thanks :-)

I just save game and load it after, i have a crash.

 i attach my savegame here + log

[gamelog.txt](https://github.com/dxx-rebirth/dxx-rebirth/files/14861762/gamelog.txt)
[player.zip](https://github.com/dxx-rebirth/dxx-rebirth/files/14861767/player.zip)

**Comment by vLKp** [comment]
`gamelog.txt` is not useful for reports of this type.  A backtrace from a core file might be useful.

The save game in `player.zip` initially appeared to be ill-formed.  Attempting to load it in D2X-Rebirth immediately dumps me back to the main screen.  On further investigation, the save game is from D1X-Rebirth.  In the future, please follow the instructions in the issue template, which prompts you for (among many other things) the game about which you are reporting an issue.

From what campaign and level was this save created?  The issue template asks about that, but you did not report that.

---

This save cannot be loaded on x86_64-pc-linux-gnu.  Initially, there were 30 walls with invalid wall indexes, each of which would be fatal.  I added a hack to reset invalid wall indexes on load.  After that, I got a crash in the renderer due to an out-of-range index trying to load a bitmap_index.

Further investigation suggests that the conversion of `segment_side_wall_tmap_write` to use `namespace serial` caused it to write in little endian, not host endian.  However, the loader was never converted, and reads data back in host endian, which produces an unwanted byte swap on big endian systems.  The fix is to switch `segment_side_wall_tmap_write` back to writing in host endian.  I dislike that fix, but changing the loader to read in little endian would break older savegames.  As is, only creation of savegames on big endian systems is broken by this bug.

The renderer issue is likely separate, and there may be other places where the save code now writes in little endian, but the loader still reads in host endian.

**Comment by BeWorld2018** [comment]
Sorry, yes D1X :-( with Descent GOG version and level 1, directly saved at startup game.

the loader still reads in host endian
 > i can confirm that, i load saved game from Windows x64 version.

**Comment by vLKp** [comment]
I believe the above commit should resolve this issue, by reverting to the previous behavior of writing this structure in native endian.  If I read the commit history right, this broke in 3714c0c64559bb1b520d500be31217d83641ef6b (October 2014).
